### DISCUSSION TITLE
grapesjs translating code after saving

### DISCUSSION THREAD

**Discussion by joanliky** [GrapesJS/grapesjs#5013] [Q&A]
I've got two question about using Grapesjs.

Q1/
I've used ckeditor as my text editor. but after being stored in the database, the attribute of the text will translate from data-gjs-type="text" to data-gjs-type="default",  and after that text can't be editable. I stored my code as HTML+CSS in database, is it possible if I can store it in HTML+CSS and my text can be still available to edit after saving? 
I search on google I've seen someone mention we can save as Json to prevent this happen, but I wonder is it possible if save in HTML+CSS?

original：
<img width="551" alt="image" src="https://user-images.githubusercontent.com/85118665/228553772-3812cf28-90c0-487e-9c02-529f5f39eb0e.png">
After save：
<img width="575" alt="image" src="https://user-images.githubusercontent.com/85118665/228512751-c2b6cf39-7c2f-4f81-bcdb-7aeba846376e.png">

 
Q2/
I had set a limit to a component about can't be copyable and removable(data-gjs-removable="false" data-gjs-copyable="false"), as the same problem i mentioned, it's been removed after save, so can i ask if it possible to keep my attribute after save?

**Comment by artf** [comment]
> I search on google I've seen someone mention we can save as Json to prevent this happen, but I wonder is it possible if save in HTML+CSS?

Yes, relying on the JSON is the proper way to update your project.
There is an option `withProps` in [`toHTML`](https://grapesjs.com/docs/api/component.html#tohtml) which allows you export grapesjs properties in your HTML but I'd say it's only valid if you're using basic components as not all the properties could be serialised.

You can make use of `optsHtml` in your initial config if you want this option to be always applied:
```js
grapesjs.init({
      optsHtml: { withProps: true },
     // ...
});
```

**Reply by joanliky** [reply]
> > I search on google I've seen someone mention we can save as Json to prevent this happen, but I wonder is it possible if save in HTML+CSS?
> 
> Yes, relying on the JSON is the proper way to update your project. There is an option `withProps` in [`toHTML`](https://grapesjs.com/docs/api/component.html#tohtml) which allows you export grapesjs properties in your HTML but I'd say it's only valid if you're using basic components as not all the properties could be serialised.
> 
> You can make use of `optsHtml` in your initial config if you want this option to be always applied:
> 
> ```js
> grapesjs.init({
>       optsHtml: { withProps: true },
>      // ...
> });
> ```


Thanks so much!! It's work on my CKeditor problem @artf !
Seems like it doesn't work on our custom button we created.

here's my button component structure:
![image](https://user-images.githubusercontent.com/85118665/229109114-674f97fd-102e-49b3-a39b-c7984dd78e30.png)

after i saved it will be turn into this:
![image](https://user-images.githubusercontent.com/85118665/229109303-7ffc1faa-5f91-443c-a848-edb90004cb37.png)

may i ask it there any possible way to fix this:(?

**Reply by artf** [reply]
Sorry but I can't see where is the issue, please elaborate more on how you create the button and how you would expect the output

**Comment by joanliky** [comment]
@artf thanks for your reply!

i've tried to stored in Json, But i've got the other problem.

my gjs version:**grapesjs - 0.20.1**

i created a block with component,  which i stored it in Json.
After i saved my newsletter and reopen it, 
button style will be remove:

Before save：
<img width="693" alt="image" src="https://user-images.githubusercontent.com/85118665/230317032-e32f590e-4d3d-4af9-98af-df47d4546bf1.png">

After save：
<img width="698" alt="image" src="https://user-images.githubusercontent.com/85118665/230316282-41e888a8-05ef-4b26-b3be-4bef4efcce02.png">

i already tried different method: style on block(which had been mention in [3872](https://github.com/GrapesJS/grapesjs/discussions/3872) )/style inline.

style on Block:
![image](https://user-images.githubusercontent.com/85118665/230314621-c5e0e505-9789-41f7-934c-ee530ebdef07.png)

style inline:
![image](https://user-images.githubusercontent.com/85118665/230518445-355ae3be-1b28-4367-8776-a8db80436315.png)

Seems like all the style will store in **gjs-css-rules**.
but the sort of style will be removed after save.

on the other hand, cause i've got outlook support needed which only support using inline VML styles.
![image](https://user-images.githubusercontent.com/85118665/230599107-74a0bc8c-a32e-4777-9688-85e0b0fc9675.png)


May i ask is there any method i can keep my block with inline style which won't be remove after save?

**Reply by joanliky** [reply]
@artf May i ask is there any method? :'(

**Reply by artf** [reply]
There might be styling issues with your version, use always the latest one.

**Reply by joanliky** [reply]
I'vr already tried updated my grapejs version.
Seem i still cant keep my inline css after saved to Json,
may i ask is it not way to keep my inline style?
